lib/internal/util.js

```diff
@@ -187,3 +191,65 @@ exports.convertToValidSignal = function convertToValidSignal(signal) {
 
   throw new Error('Unknown signal: ' + signal);
 };
+
+const kCustomPromisifiedSymbol = Symbol('util.promisify.custom');
+const kCustomPromisifyArgsSymbol = Symbol('customPromisifyArgs');
+
+function promisify(orig) {
+  if (typeof orig !== 'function') {
+    throw new TypeError('The first argument to promisify() must be a function');
+  }
+
+  if (orig[kCustomPromisifiedSymbol]) {
+    const fn = orig[kCustomPromisifiedSymbol];
+    Object.defineProperty(fn, kCustomPromisifiedSymbol, {
+      value: fn, enumerable: false, writable: false, configurable: true
+    });
+    return fn;
+  }
+
+  if (isAsyncFunction(orig)) {
```

Comment:
Come to think of it, I don't think we should special-case async functions since functions can take callbacks and return promises too and promisify should be consistent on those.

```
async function badButPossibleCode(url, cb) {
  const data = await get(url);
  callbackAPI(data, (err, data) {
   if(err) cb(err);
   else cb(null, process(data));
  });
}
```

I'd expect `util.promisify` to return a function that returns a promise for the callback'd value and not the originally returned promised.